### DISCUSSION TITLE
Add a Zotero connector to retrieve scientific bibliography

### DISCUSSION THREAD

**Discussion by flefevre** [open-webui/open-webui#6766] [Ideas]
Feature Request: Add Zotero Integration for Retrieving Scientific Papers in Knowledge Creation

# Enhancement Description: 
Introduce a new feature to allow users to connect a Zotero library when creating a "knowledge" in 'open webui'. In addition to the existing document upload functionality, add a button to directly connect to a Zotero server and retrieve a collection of scientific papers. This integration will streamline adding references and organizing research material within the tool.

# Feature Requirements:

## Zotero Integration:

Add a button labeled "Connect to Zotero" in the knowledge creation UI.

Allow users to authenticate with their Zotero account via an API key and provide access to their library.

Retrieve collections of papers from either personal or group libraries.

## Setup Requirements:

Users should install pyzotero using pip install pyzotero, if not already included.

Guide users to create a Zotero API key and set up environment variables:

ZOTERO_USER_ID: User or group library ID.

ZOTERO_API_KEY: API key with read access.

ZOTERO_URL: the server URL with port and Https

## User Parameters:

Allow users to configure the following parameters when fetching documents:

Library Type: Choose between user or group library.

Query Parameters: Enter search terms and filter criteria (e.g., keywords, document types).

Sort Options: Specify sorting preference, such as by date.

Limit: Set the maximum number of documents to retrieve.

## Document Retrieval and Filtering:

When a collection is retrieved, allow filtering based on parameters such as:

Number of pages (skip excessively long papers).

Document format (ensure that documents have PDF attachments).

Example filtering option:

if item.num_pages > 30:
    continue  # skip lengthy papers

## Adding Documents to Knowledge Base:

Integrate the Zotero items directly into the 'open webui' knowledge base:

Import PDF attachments from Zotero items.

Assign each imported document a unique identifier (e.g., Zotero item.key).

Ensure metadata (title, authors, publication date) is preserved.

## Example Code:

Here’s an example of how similar functionality is implemented in a different project using pyzotero and ZoteroDB:

Import Docs
 import ZoteroDB

docs = Docs()
zotero = ZoteroDB(library_type="user")  # "group" for group library

# Retrieve and add papers to docs
for item in zotero.iterate(limit=20):
    if item.num_pages > 30:
        continue  # skip long papers
    docs.add(item.pdf, docname=item.key)

# Specific query example
for item in zotero.iterate(
    q="large language models",
    qmode="everything",
    sort="date",
    direction="desc",
    limit=100,
):
    docs.add(item.pdf, docname=item.key)

## Acceptance Criteria:

Users can authenticate with their Zotero account and retrieve collections directly.

Retrieved documents should appear in the knowledge base with all necessary metadata.

User-specified parameters (library type, query filters, sorting, etc.) are honored during document retrieval.

The feature is stable and error handling is implemented for cases such as missing API keys, connectivity issues, and missing PDF attachment

**Comment by GJoe2** [comment]
That is a good idea, I would like to see a Zotero add-on for skipping Zotero API keys requirements

**Comment by daniel-wojahn** [comment]
I am new to the open-webui project and community, and have to say, this would be one of the most awesome features! I have fine-tuned open-webui to work as a solid workhorse in an academic context, and while the current workaround is manageable, this would makes things even easier!

**Comment by AllergicMushroom** [comment]
Would this connector be able to also create notes from the local agents in the zotero database ? That would be perfect !

**Comment by flefevre** [comment]
Now there is mcp protocol for zotero.
It is more suitable for the architecture.
I think I could close this post

https://github.com/54yyyu/zotero-mcp

**Reply by AllergicMushroom** [reply]
Indeed it seems like a suitable protocol. Out of curiosity, did you manage to use it with openwebui or have you seen somebody do it ?
